src/Dialect/ONNX/ShapeInference/Shape.cpp

```diff
@@ -9,60 +9,83 @@
 //===----------------------------------------------------------------------===//
 
 #include "src/Dialect/ONNX/ShapeInference/ONNXShapeHelper.hpp"
+#include <tuple>
 #include <utility>
 
 using namespace mlir;
 
 namespace onnx_mlir {
 
+namespace {
+
+// The Shape op spec says:
+//
+// "Note that axes will be clipped to the range [0, r-1], where r is the
+// rank of the input tensor if they are out-of-range (after adding r in the case
+// of negative axis). Thus, specifying any end value > r is equivalent to
+// specifying an end value of r, and specifying any start value < -r is
+// equivalent to specifying a start value of 0."
+int64_t normalize(int64_t axis, int64_t rank) {
```

Comment:
I see what you do here, it sounds reasonable but is it the standard? Or should we instead report an error? 
In general, we have erred in favor of reporting errors so as to make the user aware of the issue, instead of "fixing" the model the way we "hope" the user wanted.
As a minimum, I would issue a warning.
Note also that if I recall correctly, sometime the clipping is at rank and sometime rank-1... so this method may not be applied everywhere.

@chentong319 do you have an opinion on this issue?

Reply:
Sorry if this was confusing. This was actually lifted for the ONNX specification

> Shape-15
Takes a tensor as input and outputs an 1D int64 tensor containing the shape of the input tensor. Optional attributes start and end can be used to compute a slice of the input tensor's shape. If start axis is omitted, the slice starts from axis 0. The end axis, if specified, is exclusive (and the returned value will not include the size of that axis). If the end axis is omitted, the axes upto the last one will be included. Negative axes indicate counting back from the last axis. Note that axes will be clamped to the range [0, r-1], where r is the rank of the input tensor if they are out-of-range (after adding r in the case of negative axis). Thus, specifying any end value > r is equivalent to specifying an end value of r, and specifying any start value < -r is equivalent to specifying a start value of 0.

It seems weird, but I have based it off the ONNX spec of Shape.

Reply:
Great, I would then just reflect this policy in the function name, so that others may not be tempted to use it when it is not applicable. Add something of the sort "ClampedPerSpec" or something similar.